apm-agent-plugins/apm-okhttp-plugin/src/main/java/co/elastic/apm/agent/okhttp/OkHttpRequestBuilderInstrumentation.java

```diff
@@ -0,0 +1,77 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.okhttp;
+
+import co.elastic.apm.agent.bci.ElasticApmInstrumentation;
+
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.TraceContext;
+import com.squareup.okhttp.Headers;
+import com.squareup.okhttp.Request;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.implementation.bytecode.assign.Assigner;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+public class OkHttpRequestBuilderInstrumentation extends ElasticApmInstrumentation {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    private static void onBeforeExecute(@Advice.This @Nullable Object thiz,
+                                        @Advice.FieldValue(value = "headers", typing = Assigner.Typing.DYNAMIC) @Nullable Object headers) {
+        if (tracer == null || tracer.currentSpan() == null) {
+            return;
+        }
+        if (thiz == null || headers == null) {
+            return;
+        }
+        final AbstractSpan<?> currentSpan = tracer.currentSpan();
+        if (thiz instanceof Request.Builder) {
+            if (headers instanceof Headers.Builder) {
+                if (((Headers.Builder) headers).get(TraceContext.TRACE_PARENT_HEADER) == null) {
+                    ((Headers.Builder) headers).set(TraceContext.TRACE_PARENT_HEADER, currentSpan.getTraceContext().getOutgoingTraceParentHeader().toString());
```

Comment:
This probably happens before the actual span is started (in `Call.execute`), so the `parentId` part of the outgoing header would not be the one we want to send. We need some APIs we are now missing. For example, a way to create a child `TraceContext` of the active span and use it for the header creation, then make it available for the to-be-created span. We are currently missing both of these internal APIs- create a standalone context and start a span with a readymade context. Let me see think about that and see how we do that.
In any case, since we are not entirely sure there will be a span at all, we will need to make sure this does not leak.

Regardless, I just looked at `okhttp3.Request` and it seems to have a package-private constructor, so it seems to be safe to assume it is always built through the Builder. Is it the same for okhttp 2 as well? Is there another standard way to create `Request`s?